File: 0377-Combination_Sum_IV/Combination_Sum_IV.go
package combinationsumiv

func combinationSum4(nums []int, target int) int {
	dp := make([]int, target+1)
	for i := range dp {
		dp[i] = -1
	}
	return checkCombinations(nums, target, dp)
}

func checkCombinations(nums []int, target int, dp []int) int {
	if target == 0 {
		return 1
	}

	if target < 0 {
		return 0
	}

	if dp[target] != -1 {
		return dp[target]
	}

	ans := 0
	for _, num := range nums {
		ans += checkCombinations(nums, target-num, dp)
	}

	dp[target] = ans
	return ans
}

File: 0377-Combination_Sum_IV/Combination_Sum_IV.java
import java.util.*;

class Solution {
    Integer[] dp;
    public int combinationSum4(int[] nums, int target) {
        Arrays.sort(nums);
        dp = new Integer[target + 1];

        return checkSum(nums,target);
    }
    
    private int checkSum(int[] nums, int target){
        if(target == 0) return 1;
        
        if(target < 0) return 0;
        
        if(dp[target] != null) return dp[target];
        
        int ans = 0; 
        for(int i: nums){
            ans += checkSum(nums, target - i);
        }
        
        dp[target] = ans;
        return ans;
    }
}